pkg/cmd/cli/cmd/extract.go

```diff
@@ -0,0 +1,171 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	extractLong = `
+Extract files out of secrets and config maps
+
+The extract command makes it easy to download the contents of a config map or secret into a directory.
+Each key in the config map or secret are created as a separate file with the name of the key, as they
+are when you mount a secret or config map into a container.
+
+You can limit which keys are extracted with the --only=NAME flag, or set the directory to extract to
+with --to=DIRECTORY.`
+
+	extractExample = `  # extract the secret "test" to the current directory
+  %[1]s extract secret/test
+
+  # extract the config map "nginx" to the /tmp directory
+  %[1]s extract configmap/nginx --to=/tmp`
+)
+
+type ExtractOptions struct {
+	Out, Err        io.Writer
+	Filenames       []string
+	OnlyValues      []string
+	TargetDirectory string
+	Overwrite       bool
+
+	VisitorFn             func(resource.VisitorFunc) error
+	ExtractFileContentsFn func(runtime.Object) (map[string][]byte, bool, error)
+}
+
+func NewCmdExtract(fullName string, f *clientcmd.Factory, in io.Reader, out, errOut io.Writer) *cobra.Command {
+	options := &ExtractOptions{
+		Out: out,
+		Err: errOut,
+	}
+	cmd := &cobra.Command{
+		Use:     "extract RESOURCE/NAME [--to=DIRECTORY] [options]",
+		Short:   "Extract secrets or config maps to disk",
+		Long:    extractLong,
+		Example: fmt.Sprintf(extractExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(f, in, out, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			err := options.Run()
+			if err == cmdutil.ErrExit {
+				os.Exit(1)
+			}
+			kcmdutil.CheckErr(err)
+		},
+	}
+	cmd.Flags().BoolVar(&options.Overwrite, "confirm", options.Overwrite, "Overwrite files that already exist.")
+	cmd.Flags().StringVar(&options.TargetDirectory, "to", options.TargetDirectory, "Directory to extract files to.")
+	cmd.Flags().StringSliceVarP(&options.Filenames, "filename", "f", options.Filenames, "Filename, directory, or URL to file to identify to extract the resource.")
+	cmd.MarkFlagFilename("filename")
+	cmd.Flags().StringSliceVar(&options.OnlyValues, "only", options.OnlyValues, "An optional list of keys to extract (default is all keys).")
```

Comment:
Sure.  Let me look to see if we use it anywhere else.

On Wed, Jun 22, 2016 at 4:23 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/extract.go
> <https://github.com/openshift/origin/pull/9477#discussion_r68127663>:
>
> > +		Example: fmt.Sprintf(extractExample, fullName),
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			kcmdutil.CheckErr(options.Complete(f, in, out, cmd, args))
> > +			kcmdutil.CheckErr(options.Validate())
> > +			err := options.Run()
> > +			if err == cmdutil.ErrExit {
> > +				os.Exit(1)
> > +			}
> > +			kcmdutil.CheckErr(err)
> > +		},
> > +	}
> > +	cmd.Flags().BoolVar(&options.Overwrite, "confirm", options.Overwrite, "Overwrite files that already exist.")
> > +	cmd.Flags().StringVar(&options.TargetDirectory, "to", options.TargetDirectory, "Directory to extract files to.")
> > +	cmd.Flags().StringSliceVarP(&options.Filenames, "filename", "f", options.Filenames, "Filename, directory, or URL to file to identify to extract the resource.")
> > +	cmd.MarkFlagFilename("filename")
> > +	cmd.Flags().StringSliceVar(&options.OnlyValues, "only", options.OnlyValues, "An optional list of keys to extract (default is all keys).")
>
> would --key or --keys make more sense than --only?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9477/files/842b6e081e85c01eb2b964a03f6f18907f03a1e2#r68127663>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_pyCKS9h6gBc0wdj-_fQthrhpjEf0ks5qOZm3gaJpZM4I7RyR>
> .
>